src/plugins/memdump/memdump.cpp

```diff
@@ -0,0 +1,311 @@
+                    UNPACK_TIMEVAL(info->timestamp),
+                    escaped_pname,
+                    USERIDSTR(drakvuf), info->proc_data.userid,
+                    info->proc_data.pid, info->proc_data.ppid,
+                    info->trap->name, reason, info->proc_data.pid, ctx->addr,
+                    num_pages * VMI_PS_4KB, escaped_fname);
+            g_free(escaped_fname);
+            g_free(escaped_pname);
+            break;
+        default:
+        case OUTPUT_DEFAULT:
+            printf("[MEMDUMP] TIME:" FORMAT_TIMEVAL " VCPU:%" PRIu32 " CR3:0x%" PRIx64 ",\"%s\" %s:%" PRIi64" \"%s\" Reason:\"%s\" Process:%d Base:0x%" PRIx64 " Size:%" PRIu64 " File:\"%s\"\n",
+                   UNPACK_TIMEVAL(info->timestamp), info->vcpu, info->regs->cr3, info->proc_data.name,
+                   USERIDSTR(drakvuf), info->proc_data.userid, info->trap->name, reason, info->proc_data.pid, ctx->addr,
+                   num_pages * VMI_PS_4KB, file);
+            break;
+    }
+
+    printf("\n");
+
+done:
+    if (file)
+        free(file);
+
+    if (access_ptrs)
+        g_free(access_ptrs);
+
+    return ret;
+}
+
+static event_response_t free_virtual_memory_hook_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    // HANDLE ProcessHandle
+    uint64_t process_handle = drakvuf_get_function_argument(drakvuf, info, 1);
+    // OUT PVOID *BaseAddress
+    addr_t mem_base_address_ptr = drakvuf_get_function_argument(drakvuf, info, 2);
+
+    if (process_handle != 0xffffffffffffffffULL)
+    {
+        PRINT_DEBUG("[MEMDUMP] Process handle not pointing to self, ignore\n");
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    auto plugin = get_trap_plugin<memdump>(info);
+    if (!plugin)
+        return VMI_EVENT_RESPONSE_NONE;
+
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+    access_context_t ctx = { .translate_mechanism = VMI_TM_PROCESS_DTB, .dtb = info->regs->cr3 };
+    ctx.addr = mem_base_address_ptr;
+
+    addr_t mem_base_address;
+
+    if (VMI_SUCCESS != vmi_read_addr(vmi, &ctx, &mem_base_address))
+    {
+        PRINT_DEBUG("[MEMDUMP] Failed to read base address in NtFreeVirtualMemory\n");
+        drakvuf_release_vmi(drakvuf);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    mmvad_info_t mmvad;
+
+    if (VMI_SUCCESS != drakvuf_find_mmvad(drakvuf, info->proc_data.base_addr, mem_base_address, &mmvad))
+    {
+        PRINT_DEBUG("[MEMDUMP] Failed to find MMVAD for memory passed to NtFreeVirtualMemory\n");
+        drakvuf_release_vmi(drakvuf);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    ctx.addr = mem_base_address;
+    uint16_t magic;
+    char* magic_c = (char*)&magic;
+
+    if (VMI_SUCCESS != vmi_read_16(vmi, &ctx, &magic))
+    {
+        PRINT_DEBUG("[MEMDUMP] Failed to access memory to be used with NtFreeVirtualMemory\n");
+        drakvuf_release_vmi(drakvuf);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    if (magic_c[0] == 'M' && magic_c[1] == 'Z')
```

Comment:
You can avoid this variable if you just check if `magic == 0x5a4d` (not 0x4d5a because of endianness).

You can even check if `magic == 'ZM'` (because multi byte character literals), but that's pretty hacky.

Reply:
Our internal version is actually using `magic == 0x5a4d` AFAIR but this is less clear what we are checking for, so let's stay with `MZ` here.